core/vm/evm.go

```diff
@@ -838,7 +849,43 @@ func (evm *EVM) captureOperationMode(isWriteOperation bool) error {
 	return nil
 }
 
-// enforceMultitenancyCheck returns true if EVM is enforced to do multitenancy check, false otherwise
+// Quorum
+//
+// captureAffectedContract stores the contract address to the affectedContract list if not yet there.
+// The affected mode is also updated if required.
+// Default affected reason is MessageCall.
+// In simulation/eth_call for multitenancy, it sets the expectation of AffectedMode
+// to be verified later when an opcode is executed.
+func (evm *EVM) captureAffectedContract(address common.Address, mode AffectedMode) error {
+	if _, ok := evm.affectedContracts[address]; !ok {
+		evm.affectedContracts[address] = newAffectedType(MessageCall, mode)
+	}
+
+	if !evm.SupportsMultitenancy || evm.AuthorizeMessageCallFunc == nil {
+		return nil
+	}
+	affectedType, found := evm.affectedContracts[address]
+	if !found || affectedType.mode != ModeUnknown {
+		return nil
+	}
+	authorizedRead, authorizedWrite, err := evm.AuthorizeMessageCallFunc(address)
+	if err != nil {
+		return err
+	}
+	authorizedMode := affectedType.mode.Update(authorizedRead, authorizedWrite)
+	log.Debug("AffectedMode changed", "address", address.Hex(), "old", affectedType.mode, "new", authorizedMode, "read", authorizedRead, "write", authorizedWrite)
+	// if we don't authorize either read/write, it's unauthorized access
+	// and we need to inform EVM
+	if !authorizedRead && !authorizedWrite {
```

Comment:
Why not checking this right after `authorizedRead, authorizedWrite, err := evm.AuthorizeMessageCallFunc(address)` on line 871? 

Actually would it even make sense to return an error from `evm.AuthorizeMessageCallFunc()` in case both `authorizedRead` and `authorizedWrite` are `false` ?

So then we can directly do 
`affectedType.mode = affectedType.mode.Update(authorizedRead, authorizedWrite)`